src/Npgsql/NpgsqlConnection.cs

```diff
@@ -223,67 +249,83 @@ async Task OpenLong()
+                                connector = await _pool.Rent(this, timeout, async, cancellationToken);
+                                ConnectorBindingScope = ConnectorBindingScope.Connection;
+                                Connector = connector;
+                                EnlistTransaction(Transaction.Current);
                             }
                         }
-                        else  // No enlist
-                            Connector = await _pool.AllocateLong(this, timeout, async, cancellationToken);
-
-                        // Since this pooled connector was opened, types may have been added (and ReloadTypes() called),
-                        // or global mappings may have changed. Bring this up to date if needed.
-                        mapper = Connector.TypeMapper;
-                        if (mapper.ChangeCounter != TypeMapping.GlobalTypeMapper.Instance.ChangeCounter)
-                            await Connector.LoadDatabaseInfo(NpgsqlTimeout.Infinite, async);
+                        else
+                            connector = await _pool.Rent(this, timeout, async, cancellationToken);
                     }
 
-                    // We may have gotten an already enlisted pending connector above, no need to enlist in that case
-                    if (transaction != null && EnlistedTransaction == null)
-                        EnlistTransaction(Transaction.Current);
+                    ConnectorBindingScope = ConnectorBindingScope.Connection;
+                    Connector = connector;
+
+                    // Since this connector was last used, PostgreSQL types (e.g. enums) may have been added
+                    // (and ReloadTypes() called), or global mappings may have changed by the user.
+                    // Bring this up to date if needed.
+                    // Note that in multiplexing execution, the pool-wide type mapper is used so no
+                    // need to update the connector type mapper (this is why this is here).
+                    if (connector.TypeMapper.ChangeCounter != TypeMapping.GlobalTypeMapper.Instance.ChangeCounter)
+                        connector.RebindTypeMapper();
+
+                    CompleteOpen();
                 }
                 catch
                 {
-                    if (Connector != null)
+                    FullState = ConnectionState.Closed;
+                    ConnectorBindingScope = ConnectorBindingScope.None;
+                    Connector = null;
+
+                    if (connector != null)
                     {
                         if (_pool == null)
-                            Connector.Close();
+                            connector.Close();
                         else
-                            _pool.Release(Connector);
-                        Connector = null;
+                            _pool.Return(connector);
```

Comment:
If it's only about Rent failing, then I agree that any state should be cleaned up there rather than forcing the caller to Return. I think this makes some sense as a failsafe mechanism in case an exception is raised somewhere later (e.g. from RebindTypeMapper?), though I agree that strictly speaking it shouldn't be necessary.

Reply:
It's indeed a bit of a broadly scoped try catch, nothing to worry about in any case.

Reply:
Yeah, it's definitely on the defensive side.